### ISSUE TITLE
dubbo tracing support

### ISSUE THREAD

**Issue by wbpcode** [envoyproxy/envoy#17601]
After #16049 is completed, tracers can be used by various proxies, not just limited to HTTP. In envoy, now it is possible to trace dubbo/thrift streams as well as HTTP requests.

Dubbo Proxy would be a good start.

**Comment by wbpcode** [comment]
cc @zyfjeff @lizan

**Comment by yingjianjian** [comment]
> cc @zyfjeff @lizan

Hello, has Dubbo trace supported it? I'm looking forward to it @wbpcode

**Comment by wbpcode** [comment]
> > cc @zyfjeff @lizan
> 
> Hello, has Dubbo trace supported it? I'm looking forward to it @wbpcode

sorry, it's not supported for now.

**Comment by tedli** [comment]
Hi,

I'm trying to add a filter of dubbo, like the Router do, for learning purpose. Current only want to read from config and log something.

So, I created a filter, I want to read config using the `config` field of DubboFilter which is an `Any`. I give my filter a name, and register it with config field type `envoy::extensions::filters::network::http_connection_manager::v3::HttpConnectionManager::Tracing`, it built.
However when running, envoy complains about it don't know how to unmarsal yaml config to object.

```c++
DubboFilters::FilterFactoryCb LalalaFilterConfig::createFilterFactoryFromProtoTyped(
    [[maybe_unused]] const envoy::extensions::filters::network::http_connection_manager::v3::HttpConnectionManager::Tracing& proto,
    [[maybe_unused]] const std::string& stat_prefix,
    Server::Configuration::FactoryContext& context) {
    // log proto...
  return [&context](DubboFilters::FilterChainFactoryCallbacks& callbacks) -> void {
    callbacks.addFilter(std::make_shared<Lalala>(context.clusterManager()));
  };
}
```

```log
[2022-07-26 17:01:19.835][9004][critical][main] [source/server/server.cc:116] error initializing configuration '/path/envoy.yaml': Unable to parse JSON as proto (INVALID_ARGUMENT:(dubbo_filters[1].config): invalid value Invalid type URL, unknown type: extensions.filters.network.http_connection_manager.v3.HttpConnectionManager.Tracing for type Any)
```

Is there some docs about this? Or there already some other filters, codes that did things like this, that I can imitate？

**Comment by tedli** [comment]
Hi @wbpcode ,

If I want to achieve tracing for dubbo_proxy, for the filter's config,  I can just reuse the `Tracing` proto defined in http connection manager, rather than introduce a new proto.

Did I get it wrong?

**Comment by wbpcode** [comment]
> Hi @wbpcode ,
> 
> If I want to achieve tracing for dubbo_proxy, for the filter's config, I can just reuse the `Tracing` proto defined in http connection manager, rather than introduce a new proto.
> 
> Did I get it wrong?

I think it'ok to reuse the exist `Tracing` proto for now. May be we will have protocol-independent proto for tracing in the future, but it will be something just like current `Tracing` proto.

**Comment by tedli** [comment]
Hi @wbpcode ,

https://github.com/envoyproxy/envoy/blob/5313fa4206217e831367932a21c3e3bc6bf6efb4/source/extensions/filters/network/http_connection_manager/config.cc#L166
https://github.com/envoyproxy/envoy/blob/5313fa4206217e831367932a21c3e3bc6bf6efb4/source/extensions/filters/network/http_connection_manager/config.cc#L190-L195

I want to use the factory to create tracer implementation object like the http connection manager do. It uses singleton mechanism.

Should I import this in dubbo proxy, to reuse the singleton http_tracer_manager instance, or to imitate it, give a `dubbo_tracer_manager` like this:

```c++
DubboFilters::FilterFactoryCb LalalaFilterConfig::createFilterFactoryFromProtoTyped(
    const envoy::extensions::filters::network::http_connection_manager::v3::HttpConnectionManager::
        Tracing& proto,
    const std::string&, Server::Configuration::FactoryContext& context) {

  Tracing::HttpTracerSharedPtr tracer = nullptr;

  if (proto.has_provider()) {
    Utility::Singletons singletons = Utility::createSingletons(context);
    tracer = singletons.dubbo_tracer_manager_->getOrCreateHttpTracer(&proto.provider());
  }

  return [&context, tracer](DubboFilters::FilterChainFactoryCallbacks& callbacks) -> void {
    callbacks.addFilter(std::make_shared<Lalala>(context.clusterManager(), tracer));
  };
}

SINGLETON_MANAGER_REGISTRATION(dubbo_tracer_manager);

Utility::Singletons Utility::createSingletons(Server::Configuration::FactoryContext& context) {
  auto dubbo_tracer_manager = context.singletonManager().getTyped<Tracing::HttpTracerManagerImpl>(
      SINGLETON_MANAGER_REGISTERED_NAME(dubbo_tracer_manager), [&context] {
        return std::make_shared<Tracing::HttpTracerManagerImpl>(
            std::make_unique<Tracing::TracerFactoryContextImpl>(
                context.getServerFactoryContext(), context.messageValidationVisitor()));
      });
  return {dubbo_tracer_manager};
}
```

If import and reuse, the trade off is initialized some other object like the `route_config_provider_manager`, which is useless in dubbo_proxy. (the side effect in the http ver `createSingletons()`). Or if create a dubbo_tracer_manager, it go against the 'singleton' purpose.

Need some suggestion. Thanks.

**Comment by wbpcode** [comment]
Basically, this could be achieved by the new generic proxy-based dubbo proxy. So, I will close it. If someone still want this, ping me to reopen it or create a new issue.

**Comment by cwww3** [comment]
has Dubbo trace supported now?

**Comment by wbpcode** [comment]
cc @cwww3 sorry, but I won't add new feature to `dubbo_proxy` for now because the limited free bandwidth.

IMO, it's recommend to use the new `generic_proxy` with dubbo codec to proxy dubbo traffic. And it provide full featured tracing, logging, and stats support.

See https://www.envoyproxy.io/docs/envoy/latest/configuration/listeners/network_filters/generic_proxy_filter
